### PR TITLE
ArPow stage 1: local source-build infrastructure

### PR DESCRIPTION
**Pull request by MichaelSimons**
This PR adds the local build infrastructure that lets ArPow (arcade-powered source-build) run in this repo. See <https://github.com/dotnet/source-build/blob/master/Documentation/planning/arcade-powered-source-build/onboarding/local-onboarding.md> for more details about how it works.

To try it out locally, run this on Linux: `./build.sh -c Release --restore --build --pack /p:ArcadeBuildFromSource=true -bl`

This PR should have no effect on ordinary builds, or CI. ArPow stage 2 will add source-build to CI: PR validation and official builds.

For <https://github.com/dotnet/source-build/blob/master/Documentation/planning/arcade-powered-source-build/implementation-plan.md>

Related to #https://github.com/dotnet/source-build/issues/2071

This PR is currently blocked by the [work to bring in Arcade 6.0](https://github.com/dotnet/roslyn/pull/51647)

### CODE REVIEW

File: eng/targets/Imports.targets
Comment (MichaelSimons) [inline eng/targets/Imports.targets:179]: This issue should get addressed before merging.

**Comment by 333fred** [comment]
> This PR should have no effect on ordinary builds, or CI. ArPow stage 2 will add source-build to CI: PR validation and official builds.

We already have a CI leg for source-build, I think we should update that in the same PR. It's this section of the YAML: https://github.com/dotnet/roslyn/blob/main/azure-pipelines.yml#L201-L223.

**Comment by 333fred** [comment]
@MichaelSimons where you actually expecting a review at this stage?

**Comment by MichaelSimons** [comment]
> were you actually expecting a review at this stage?

I'm not expecting a review until the Arcade 6.0 blocking issue is resolved.  That being said I would appreciate any initial comments such the CI one @333fred made.

**Comment by MichaelSimons** [comment]
@dseefeld, @333fred - this is ready for review now.

**Comment by 333fred** [comment]
@MichaelSimons there are a large number of patches here. Why so many? At one point in the 2.2-3.0 timeframe we had removed all the patches from source-build for roslyn, but we appear to have to have added more back since then, for things that really could and should have been fixed in source directly rather than having source-build silently add patches without telling us. Some I don't really understand at all, such as the one that adds `IsExternalInit`: the compiler would not build without this type defined. I honestly don't feel particularly comfortable adding so many patch files that touch so much of the infrastructure directly, even if the goal is to immediately remove them.

File: eng/source-build-patches/0017-Rename-IReadOnlySet-and-ReferenceEqualityComparer-Ro.patch
Comment (333fred) [inline eng/source-build-patches/0017-Rename-IReadOnlySet-and-ReferenceEqualityComparer-Ro.patch:95]: This touches too many files to realistically maintain for any amount of time as a patch.

File: eng/source-build-patches/0013-Add-file-to-include-missing-IsExternalInit-class.patch
Comment (333fred) [inline eng/source-build-patches/0013-Add-file-to-include-missing-IsExternalInit-class.patch:18]: This should already be defined at https://github.com/dotnet/roslyn/blob/main/src/Compilers/Core/Portable/InternalUtilities/IsExternalInit.cs. If that isn't getting included in source-build, that's likely a problem in an of itself.

File: eng/source-build-patches/0001-Conditionally-remove-net472-from-TargetFrameworks.patch
Comment (333fred) [inline eng/source-build-patches/0001-Conditionally-remove-net472-from-TargetFrameworks.patch:18]: I don't really understand the point of these tiny patches vs just updating the source right now. It would be simpler to review, have less chance of messing up particular legs, and have less maintenance burden. I'd much rather we just make the change in one step.

File: eng/source-build-patches/0006-Dirty-revert-of-https-github.com-dotnet-roslyn-issue.patch
Comment (333fred) [inline eng/source-build-patches/0006-Dirty-revert-of-https-github.com-dotnet-roslyn-issue.patch:5]: Doesn't have to be resolved right now, but @crummel what is this doing? This seems like it introduces a bug?

File: eng/source-build-patches/0012-Build-Roslyn-as-net6.0-in-source-build.patch
Comment (333fred) [inline eng/source-build-patches/0012-Build-Roslyn-as-net6.0-in-source-build.patch:4]: @jaredpar this patch may end up being a bit troublesome. Not sure what you think.

File: src/Compilers/Core/Portable/InternalUtilities/PlatformAttributes.cs
Comment (333fred) [inline src/Compilers/Core/Portable/InternalUtilities/PlatformAttributes.cs:7]:
```suggestion
#if !NET5_0_OR_GREATER
```

File: eng/build.sh
Comment (333fred) [inline eng/build.sh:309]: Does this cause arcade to set the `DotNetBuildFromSource` environment variable?

**Review by 333fred — COMMENTED** [review]
One suggestion and one question before merging.

File: src/Compilers/Core/Portable/InternalUtilities/PlatformAttributes.cs
Comment (MichaelSimons) [reply src/Compilers/Core/Portable/InternalUtilities/PlatformAttributes.cs]: Fixed.

File: eng/build.sh
Comment (MichaelSimons) [reply eng/build.sh:309]: Yes it does.

File: eng/source-build-patches/0013-Add-file-to-include-missing-IsExternalInit-class.patch
Comment (MichaelSimons) [reply eng/source-build-patches/0013-Add-file-to-include-missing-IsExternalInit-class.patch]: I verified this is patch is not necessary and removed it.

**Comment by MichaelSimons** [comment]
> @MichaelSimons there are a large number of patches here. Why so many? At one point in the 2.2-3.0 timeframe we had removed all the patches from source-build for roslyn, but we appear to have to have added more back since then, for things that really could and should have been fixed in source directly rather than having source-build silently add patches without telling us.

We (source-build devs and Fred from Roslyn) met on this.  The pain and frustration of the patch maintenance and quality is one of the major reasons for the ArPow effort.  The current source-build strategy has many flaws/downsides.  After the discussion, a decision was made to use a feature branch to commit the ArPow infra and then work to remove the patches before merging to main.  Source-build folks will be responsible for the ArPow infra and Roslyn folks will be responsible for the patch removal with assistance from source-build folks on understanding why patches are needed.

File: eng/build.sh
Comment (MichaelSimons) [reply eng/build.sh:309]: `DotNetBuildFromSource` is still the appropriate variable to use when conditionalizing for source-build.

File: eng/source-build-patches/0012-Build-Roslyn-as-net6.0-in-source-build.patch
Comment (jaredpar) [reply eng/source-build-patches/0012-Build-Roslyn-as-net6.0-in-source-build.patch:4]: Yeah this patch is problematic. It's essentially forcing us to support running on `netcoreapp3.1` even though our code is developed with the assumption that this is just not possible. 

What is the reasoning behind this? If we actually need to support `netcoreapp3.1` then we may have to accept that. But even so I'd just add it back into the main tree. As long as it exists as a patch like this we're just going to be breaking support left and right here. It's not sustainable.

**Review by 333fred — APPROVED** [review]
I'm going to go ahead and merge this into the feature branch for now, and we'll work on getting the patches actually merged into the repo. @MichaelSimons and @crummel, there are a couple of questions that need to be answered as well, but I'm not going to hold up the merge since we'll need to address them in follow ups anyway.

**Pull request merged**
